File: solutions/python/atbash-cipher/1/atbash_cipher.py
def mirror_char(c: str) -> str:
    return chr(25 - ord(c) + 2 * ord("a"))


def encode(plain_text: str) -> str:
    result = ""
    count = 0
    for c in plain_text:
        if not c.isalnum():
            continue
        if count and count % 5 == 0:
            result += " "
        c = c.lower()
        result += mirror_char(c) if c.isalpha() else c
        count += 1
    return result


def decode(ciphered_text: str) -> str:
    return "".join(
        mirror_char(c) if c.isalpha() else c
        for c in ciphered_text
        if c.isalnum()
    )

File: solutions/c/atbash-cipher/1/atbash_cipher.c
#include "atbash_cipher.h"

#include <ctype.h>
#include <stdlib.h>
#include <string.h>

static char mirror_char(char c) { return 219 - c; }

char* atbash_encode(const char* input) {
    char* out = malloc((strlen(input) / 5 + 1) * 6 + 1);
    size_t out_len = 0;
    size_t count = 0;
    for (const char* c = input; *c; ++c) {
        if (!isalnum(*c)) continue;
        if (count && count % 5 == 0) out[out_len++] = ' ';
        out[out_len++] = isalpha(*c) ? mirror_char(tolower(*c)) : *c;
        count++;
    }
    out[out_len] = 0;
    return out;
}

char* atbash_decode(const char* input) {
    char* out = malloc(strlen(input) + 1);
    size_t out_len = 0;
    for (const char* c = input; *c; ++c)
        if (isalnum(*c))
            out[out_len++] = isalpha(*c) ? mirror_char(tolower(*c)) : *c;
    out[out_len] = 0;
    return out;
}
